### PR TITLE
CalculateAverage_michaljonko

### PR DESCRIPTION
**Pull request by michaljonko**
#### Check List:
- [x] Tests pass (`./test.sh <username>` shows no differences between expected and actual outputs)
- [x] All formatting changes by the build are committed
- [x] Your launch script is named `calculate_average_<username>.sh` (make sure to match casing of your GH user name) and is executable
- [x] Output matches that of `calculate_average_baseline.sh`
- [x] For new entries, or after substantial changes: When implementing custom hash structures, please point to where you deal with hash collisions (line number) - _Line 114_

* Execution time: ~2,9s
* Execution time of reference implementation: ~2m
* Machine: Intel i9-12900HK, 64GB Ram, PM9A1 NVMe

### CODE REVIEW

File: src/main/java/dev/morling/onebrc/CalculateAverage_michaljonko.java [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+                stationNameHash = 31 * stationNameHash + b;
+            }
+
+            while (memorySegmentOffset < memorySegmentSize && (b = memorySegment.get(ValueLayout.JAVA_BYTE, memorySegmentOffset++)) != '\n') {
+                temperatureRaw[temperatureRawIndex++] = b;
+            }
+
+            final var temperature = TemperatureParser.parse(temperatureRaw, temperatureRawIndex);
+            stationMeasurement = stationsMap.get(stationNameHash);
+            if (stationMeasurement != null) {
```
Comment (japplis) [inline src/main/java/dev/morling/onebrc/CalculateAverage_michaljonko.java:115]: I think that if 2 different station names would have the same hash, the first in the map would get the temperature of the other station.

Comment (gunnarmorling) [reply src/main/java/dev/morling/onebrc/CalculateAverage_michaljonko.java:115]: Hum, yeah, indeed. You can't condense a 100 bytes string to an integer in a map without losing information.

**Review by gunnarmorling — CHANGES_REQUESTED** [review]
Seems prone to hash collisions.

File: src/main/java/dev/morling/onebrc/CalculateAverage_michaljonko.java (hunk 1 above)
Comment (michaljonko) [reply src/main/java/dev/morling/onebrc/CalculateAverage_michaljonko.java:120]: For a case with known 10k cities (`weather_stations.csv`) it works. The last commits introduced a solution to minimize collisions when 10000 different city names are used.

**Comment by michaljonko** [comment]
@gunnarmorling do you agree with the new solution?

**Comment by gunnarmorling** [comment]
Unless I miss something, it still isn't guaranteed that all hash collisions are resolved, right? A valid entry must work correctly for _all_ valid station names.

**Comment by gunnarmorling** [comment]
Gonna close this one for now as I haven't heard back and need to clean the slate a bit. Please feel free to re-open when you feel the time has come. Thx!

**Pull request closed without merging**